sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/utils/CalciteUtils.java

```diff
@@ -189,4 +192,19 @@ private static RelDataType toRelDataType(
 
     return dataTypeFactory.createTypeWithNullability(type, field.getNullable());
   }
+
+  /**
+   * SQL-Java type mapping, with specified Beam rules: <br>
+   * 1. redirect {@link ReadableInstant} to {@link Date} so Calcite can recognize it.
+   *
+   * @param rawType
+   * @return
+   */
+  public static RelDataType sqlTypeWithAutoCast(RelDataTypeFactory typeFactory, Type rawType) {
+    //For Joda time types, return SQL type for java.util.Date.
+    if (rawType instanceof Class && ReadableInstant.class.isAssignableFrom((Class<?>) rawType)) {
```

Comment:
I think we use just `Instant` everywhere in Beam?

Reply:
`org.apache.beam.sdk.values.Row` support `AbstractInstant` as value. It's very possible that someone use other types, like `org.joda.time.DateTime`.

Reply:
Don't we want to support `DateTime` as well?

Reply:
If `Datetime` is Joda's `Datetime`, that `Datetime` implements `ReadableInstant` interface already: http://joda-time.sourceforge.net/apidocs/org/joda/time/DateTime.html

Reply:
Yes, `org.joda.time.DateTime` is supported here.

Reply:
Ah yes, but then I still don' understand why not use `Instant` or even `AbstractInstant`.

Reply:
Sorry that I didn't realize this would be a potential block question. `ReadableInstant` is used here mostly because that's the constrain in `Row` verification part:
```java
    private Instant verifyDateTime(Object value, String fieldName) {
      // We support the following classes for datetimes.
      if (value instanceof AbstractInstant) {
        return ((AbstractInstant) value).toInstant();
      } else {
        throw new IllegalArgumentException(
            String.format(
                "For field name %s and DATETIME type got unexpected class %s ",
                fieldName, value.getClass()));
      }
    }
```

That means developers can assign an `AbstractInstant` value to a DATETIME field. IMO it's reasonable to accept `AbstractInstant` as input/output type in UDF.

Reply:
No, it's not a blocker, sorry for nitpicking, I'm just trying to understand for myself, and I am still missing something :)

1. `Rows` can only contain `AbstractInstant` in `DATETIME` fields;
2. `AbstractInstant` is a class that implements `ReadableInstant` interface;
3. Other classes that do not inherit from `AbstractInstant` can potentially also implement `ReadableInstant`;
4. But we don't support other classes, at runtime we can only get `AbstractInstants` as UDF arguments because of validation in `Row`;
5. We don't rely on anything specific to `ReadableInstant`;

So why not use the same `AbstractInstant` as in `Row`?

Reply:
Good points here, seems it's the inconsistent between Row` and `BeamSqlPrimitive`:
1. `Row` accepts `AbstractInstant` and always store/return an `Instant`;
2. `BeamSqlPrimitive` accepts `ReadableInstant` and return as it is;

create https://issues.apache.org/jira/browse/BEAM-5976 to fix it.